### PR TITLE
Choose ocrd workflow

### PR DESCRIPTION
**Pull request by markusweigelt**
Add ocr workflow selection of workflows in "ocr_workflows" subfolder of config modules. If selected file will be copied to process directory with name ocr_workflow.

### CODE REVIEW

File: Kitodo/src/main/java/org/kitodo/production/services/ocr/OCRWorkflowService.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import org.primefaces.model.SortOrder;
+
+public class OCRWorkflowService extends SearchDatabaseService<OCRWorkflow, OCRWorkflowDAO> {
+
+    private static volatile OCRWorkflowService instance = null;
+
+    /**
+     * Constructor necessary to use searcher in child classes.
+     */
+    public OCRWorkflowService() {
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/production/services/ocr/OCRWorkflowService.java:33]: Should this method not be ` private` as a public class constructor ruins the ` getInstance()` usage?

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/database/persistence/OCRWorkflowDAO.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+import org.kitodo.data.database.beans.OCRWorkflow;
+import org.kitodo.data.database.exceptions.DAOException;
+
+public class OCRWorkflowDAO extends BaseDAO<OCRWorkflow> {
+
+    @Override
+    public OCRWorkflow getById(Integer id) throws DAOException {
+        OCRWorkflow ocrWorkflow = retrieveObject(OCRWorkflow.class, id);
+        if (ocrWorkflow == null) {
```
Comment (henning-gerhardt) [inline Kitodo-DataManagement/src/main/java/org/kitodo/data/database/persistence/OCRWorkflowDAO.java:25]: `Objects.isNull()` call is used every else.

File: Kitodo/src/main/java/org/kitodo/production/forms/OCRWorkflowEditView.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /**
+     * Get list of ocr workflow filenames.
+     *
+     * @return list of ocr workflow filenames
+     */
+    public List<Path> getOCRWorkflowFilenames() {
+        try (Stream<Path> ocrWorkflowPaths = Files.walk(Paths.get(ConfigCore.getParameter(ParameterCore.DIR_OCR_WORKFLOWS)))) {
+            return ocrWorkflowPaths.filter(f -> f.toString().endsWith(".sh"))
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/production/forms/OCRWorkflowEditView.java:91]: What will happen here on non Unix/Linux systems with a non ending .sh file extension? On Windows `.bat` is common name for such kind of script.

File: Kitodo/src/main/java/org/kitodo/config/enums/ParameterCore.java [hunk 4]
```diff
@@ -46,6 +46,12 @@ public enum ParameterCore implements ParameterInterface {
      */
     DIR_RULESETS(new Parameter<UndefinedParameter>("directory.rulesets")),
 
+    /**
+     * Absolute path to the directory that the ocr workflow files will be
+     * read from. It must be terminated by a directory separator ("/").
+     */
+    DIR_OCR_WORKFLOWS(new Parameter<UndefinedParameter>("directory.ocr.workflows")),
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/config/enums/ParameterCore.java:53]: You are defining here a new ` kitodo_config.properties` entry but did not provide an adjusted configuration file nor a documentation for this kind of entry.

File: Kitodo-DataManagement/src/main/resources/db/migration/V2_109__Add_authorities_to_manage_ocr_workflow.sql [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    PRIMARY KEY(id)
+) DEFAULT CHARACTER SET = utf8mb4
+  COLLATE utf8mb4_unicode_ci;
+
+-- Add column related to ocr workflow to template table
+ALTER TABLE template ADD ocr_workflow_id INT(11) DEFAULT NULL;
+
+-- Add foreign key
+ALTER TABLE template add constraint `FK_template_ocr_workflow_id`
+    foreign key (ocr_workflow_id) REFERENCES ocrworkflow(id);
```
Comment (henning-gerhardt) [inline Kitodo-DataManagement/src/main/resources/db/migration/V2_109__Add_authorities_to_manage_ocr_workflow.sql:34]: Missing new line at end of file.

**Review by henning-gerhardt — COMMENTED** [review]
Beside my comments - not a single line of test :-( But how needs automatic executed tests?

**Comment by Kathrin-Huber** [comment]
please reopen, if it is not a draft anymore

**Pull request closed without merging**